File: kernel/fdt.cc
// kernel/fdt.cc — Minimal FDT (Device Tree Blob) scanner for ARM64
//
// Walks the DTB structure block looking for nodes with specific compatible
// strings and extracts their MMIO base addresses from the reg property.
//
// Assumptions (standard for QEMU virt and Apple VZ):
//   • #address-cells = 2, #size-cells = 2 at root level
//   • reg addresses stored as (hi32 BE, lo32 BE) → 64-bit address
//   • Devices of interest are direct children of root (depth 1) or /soc (depth
//   2)

#include "kernel/fdt.h"

#if defined(__aarch64__)

namespace fdt {

static Info g_info;

// ---- Big-endian helpers ----------------------------------------------------

static inline uint32_t be32(const uint8_t *p) {
  return ((uint32_t)p[0] << 24) | ((uint32_t)p[1] << 16) |
         ((uint32_t)p[2] << 8) | (uint32_t)p[3];
}

// Read a 2-cell big-endian address (hi32, lo32) → 64-bit.
static inline uint64_t be64_2cell(const uint8_t *p) {
  return ((uint64_t)be32(p) << 32) | be32(p + 4);
}

// ---- String helpers --------------------------------------------------------

static bool str_eq(const char *a, const char *b) {
  while (*a && *b) {
    if (*a++ != *b++)
      return false;
  }
  return *a == *b;
}

// Check if the null-string list [data, data+len) contains null-terminated
// 'needle'.
static bool strlist_has(const uint8_t *data, uint32_t len, const char *needle) {
  const uint8_t *end = data + len;
  const uint8_t *p = data;
  while (p < end) {
    if (str_eq((const char *)p, needle))
      return true;
    while (p < end && *p)
      ++p;
    if (p < end)
      ++p; // skip null terminator
  }
  return false;
}

// ---- DTB structure tokens (all big-endian uint32) --------------------------

static constexpr uint32_t FDT_MAGIC = 0xD00DFEED;
static constexpr uint32_t FDT_BEGIN_NODE = 1;
static constexpr uint32_t FDT_END_NODE = 2;
static constexpr uint32_t FDT_PROP = 3;
static constexpr uint32_t FDT_NOP = 4;
static constexpr uint32_t FDT_END = 9;

// Compatible flags set while processing a single node's properties.
static constexpr int CF_PL011 = 1 << 0;  // "arm,pl011"
static constexpr int CF_VIRTIO = 1 << 1; // "virtio,mmio"
static constexpr int CF_GICV2 = 1 << 2;  // "arm,gic-400" / "arm,cortex-a15-gic"
static constexpr int CF_PCI =
    1 << 3; // device_type="pci" or "pci-host-ecam-generic"
static constexpr int CF_MEMORY =
    1 << 4;                            // device_type="memory" — physical RAM range
static constexpr int CF_GICV3 = 1 << 5; // "arm,gic-v3"

// ---- Main parser -----------------------------------------------------------

void init(uint64_t dtb_addr) {
  if (dtb_addr == 0)
    return;
  const uint8_t *dtb = reinterpret_cast<const uint8_t *>(dtb_addr);

  if (be32(dtb) != FDT_MAGIC)
    return; // not a valid DTB

  const uint32_t off_struct = be32(dtb + 8);
  const uint32_t off_strings = be32(dtb + 12);

  const uint8_t *strings = dtb + off_strings;
  const uint8_t *p = dtb + off_struct;

  // Per-node state stack (depth-indexed).
  // We scan nodes at any depth — devices can be under /soc or root.
  static constexpr int MAX_DEPTH = 8;
  struct NodeState {
    int compat;                 // CF_* flags set from "compatible" property
    uint64_t reg;               // first address from "reg" property
    uint64_t reg_size;          // first size from "reg" property
    bool has_reg;               // true once "reg" has been read
    uint64_t reg2;              // second address (e.g. GICv3 GICR base)
    bool has_reg2;              // true if second reg entry exists
    const uint8_t *ranges_data; // pointer into DTB for "ranges" property value
    uint32_t ranges_len;        // byte length of "ranges" value
    uint32_t irq;               // GIC IRQ ID parsed from "interrupts" property
    bool has_irq;               // true once "interrupts" has been read
    const uint8_t *int_map_data; // "interrupt-map" property data
    uint32_t int_map_len;        // "interrupt-map" property length
  };
  NodeState stack[MAX_DEPTH] = {};
  int depth = 0;

  while (true) {
    // Each token is a 4-byte big-endian word.
    uint32_t tok = be32(p);
    p += 4;

    switch (tok) {

    case FDT_BEGIN_NODE: {
      // Push a fresh node state for the new depth level.
      if (depth < MAX_DEPTH)
        stack[depth] = {};
      depth++;
      // Skip the null-terminated node name.
      while (*p)
        ++p;
      ++p; // skip the null byte
      // Align to 4-byte boundary.
      p = reinterpret_cast<const uint8_t *>(
          (reinterpret_cast<uintptr_t>(p) + 3) & ~(uintptr_t)3);
      break;
    }

    case FDT_END_NODE: {
      // Process the node we're exiting (stack[depth-1]).
      depth--;
      if (depth >= 0 && depth < MAX_DEPTH) {
        const NodeState &ns = stack[depth];
        if (ns.has_reg) {
          if ((ns.compat & CF_PL011) && g_info.uart_base == 0)
            g_info.uart_base = ns.reg;

          if ((ns.compat & CF_VIRTIO) && g_info.virtio_count < 32) {
            int idx = g_info.virtio_count;
            g_info.virtio_bases[idx] = ns.reg;
            g_info.virtio_irqs[idx] = ns.has_irq ? ns.irq : 0;
            g_info.virtio_count++;
          }

          if ((ns.compat & CF_GICV2) && g_info.gic_dist_base == 0) {
            g_info.gic_dist_base = ns.reg;
            g_info.gic_version = 2;
          }

          if ((ns.compat & CF_GICV3) && g_info.gic_dist_base == 0) {
            g_info.gic_dist_base = ns.reg;
            g_info.gic_redist_base = ns.has_reg2 ? ns.reg2 : 0;
            g_info.gic_version = 3;
          }

          if ((ns.compat & CF_PCI) && g_info.pcie_ecam_base == 0) {
            g_info.pcie_ecam_base = ns.reg;
            g_info.pcie_ecam_size = ns.reg_size;
          }

          // Parse PCI "ranges" to find the 32-bit MMIO aperture.
          // Each entry: [flags(4), child_hi(4), child_lo(4),
          //              cpu_hi(4), cpu_lo(4), size_hi(4), size_lo(4)] = 28
          //              bytes.
          // Space type bits[25:24]: 0=cfg 1=I/O 2=32-bit-mem 3=64-bit-mem.
          if ((ns.compat & CF_PCI) && ns.ranges_data && ns.ranges_len >= 28 &&
              g_info.pci_mmio32_base == 0) {
            for (uint32_t off = 0; off + 28 <= ns.ranges_len; off += 28) {
              const uint8_t *e = ns.ranges_data + off;
              uint32_t flags = be32(e);
              int space = (flags >> 24) & 3;
              if (space == 2) { // 32-bit memory space
                uint64_t cpu_hi = be32(e + 12), cpu_lo = be32(e + 16);
                g_info.pci_mmio32_base = (cpu_hi << 32) | cpu_lo;
                g_info.pci_mmio32_size =
                    ((uint64_t)be32(e + 20) << 32) | be32(e + 24);
                break;
              }
            }
          }

          // Parse PCI interrupt-map: maps (slot, pin) → GIC INTID.
          // Entry layout (10 cells = 40 bytes):
          //   child_addr(3) + child_int(1) + parent_phandle(1)
          //   + parent_addr(2) + parent_int(3 = type, irq_num, flags).
          if ((ns.compat & CF_PCI) && ns.int_map_data &&
              ns.int_map_len >= 40) {
            for (uint32_t off = 0; off + 40 <= ns.int_map_len; off += 40) {
              const uint8_t *e = ns.int_map_data + off;
              uint32_t child_hi = be32(e);      // PCI address hi
              // e+12: INTx pin (1=A..4=D), not needed for slot lookup
              uint32_t gic_type = be32(e + 28); // 0=SPI, 1=PPI
              uint32_t gic_irq = be32(e + 32);   // GIC irq_num
              uint32_t slot = (child_hi >> 11) & 0x1F;
              uint32_t intid =
                  (gic_type == 0) ? gic_irq + 32 : gic_irq + 16;
              // Store first interrupt mapping for each slot.
              // PCI DT may use 1-based (1=INTA) or 0-based pin numbering.
              if (slot < 8 && g_info.pci_irqs[slot] == 0)
                g_info.pci_irqs[slot] = intid;
            }
          }

          if ((ns.compat & CF_MEMORY) && g_info.ram_size == 0) {
            g_info.ram_base = ns.reg;
            g_info.ram_size = ns.reg_size;
          }
        }
      }
      break;
    }

    case FDT_PROP: {
      uint32_t vlen = be32(p);
      p += 4;
      uint32_t nameoff = be32(p);
      p += 4;
      const uint8_t *vdata = p;
      // Advance past the value (padded to 4-byte alignment).
      p += (vlen + 3) & ~(uint32_t)3;

      // Process property for the current node (stack[depth-1]).
      int d = depth - 1;
      if (d >= 0 && d < MAX_DEPTH) {
        const char *pname = (const char *)(strings + nameoff);

        if (str_eq(pname, "compatible")) {
          if (strlist_has(vdata, vlen, "arm,pl011"))
            stack[d].compat |= CF_PL011;
          if (strlist_has(vdata, vlen, "virtio,mmio"))
            stack[d].compat |= CF_VIRTIO;
          if (strlist_has(vdata, vlen, "arm,gic-400") ||
              strlist_has(vdata, vlen, "arm,cortex-a15-gic"))
            stack[d].compat |= CF_GICV2;
          if (strlist_has(vdata, vlen, "pci-host-ecam-generic"))
            stack[d].compat |= CF_PCI;
          if (strlist_has(vdata, vlen, "arm,gic-v3"))
            stack[d].compat |= CF_GICV3;
        } else if (str_eq(pname, "device_type")) {
          // Canonical way to identify PCIe host bridge nodes.
          if (strlist_has(vdata, vlen, "pci"))
            stack[d].compat |= CF_PCI;
          // Physical RAM nodes.
          if (strlist_has(vdata, vlen, "memory"))
            stack[d].compat |= CF_MEMORY;
        } else if (str_eq(pname, "reg") && vlen >= 8 && !stack[d].has_reg) {
          // First address from reg: two 32-bit big-endian cells.
          stack[d].reg = be64_2cell(vdata);
          // Read size if present (next 2 cells after address).
          stack[d].reg_size = (vlen >= 16) ? be64_2cell(vdata + 8) : 0;
          stack[d].has_reg = true;
          // Second reg entry (e.g. GICv3 redistributor base).
          // Each entry is 16 bytes (#address-cells=2, #size-cells=2).
          if (vlen >= 32) {
            stack[d].reg2 = be64_2cell(vdata + 16);
            stack[d].has_reg2 = true;
          }
        } else if (str_eq(pname, "ranges") && vlen > 0) {
          stack[d].ranges_data = vdata;
          stack[d].ranges_len = vlen;
        } else if (str_eq(pname, "interrupt-map")) {
          stack[d].int_map_data = vdata;
          stack[d].int_map_len = vlen;
        } else if (str_eq(pname, "interrupts") && vlen >= 12 &&
                   !stack[d].has_irq) {
          // GIC interrupt specifier: <type irq_num flags> (3 cells).
          // type 0 = SPI (GIC IRQ ID = irq_num + 32).
          // type 1 = PPI (GIC IRQ ID = irq_num + 16).
          uint32_t type = be32(vdata);
          uint32_t irq_num = be32(vdata + 4);
          stack[d].irq = (type == 0) ? irq_num + 32 : irq_num + 16;
          stack[d].has_irq = true;
        }
      }
      break;
    }

    case FDT_NOP:
      break;

    case FDT_END:
      return; // clean exit

    default:
      return; // corrupted DTB — bail out
    }
  }
}

const Info &info() { return g_info; }

} // namespace fdt

#endif // __aarch64__

File: kernel/fdt.rs
// kernel/fdt.rs — Minimal FDT (Device Tree Blob) scanner for ARM64
//
// Walks the DTB structure block looking for nodes with specific compatible
// strings and extracts their MMIO base addresses from the reg property.
//
// Assumptions (standard for QEMU virt and Apple VZ):
//   - #address-cells = 2, #size-cells = 2 at root level
//   - reg addresses stored as (hi32 BE, lo32 BE) -> 64-bit address
//   - Devices of interest are direct children of root (depth 1) or /soc
//
// On x86_64 this module provides no-op stubs.

use core::ptr;

// ============================================================================
// FDT Info struct — shared with C++ callers and Rust crates
// ============================================================================

/// Device information discovered from the FDT.
/// Layout must match kernel/aarch64/fdt.h fdt::Info exactly.
#[repr(C)]
pub struct FdtInfo {
    pub uart_base: u64,
    pub virtio_bases: [u64; 32],
    pub virtio_irqs: [u32; 32],
    pub virtio_count: i32,
    pub gic_dist_base: u64,
    pub gic_redist_base: u64,
    pub gic_version: u8,
    pub pcie_ecam_base: u64,
    pub pcie_ecam_size: u64,
    pub ram_base: u64,
    pub ram_size: u64,
    pub pci_mmio32_base: u64,
    pub pci_mmio32_size: u64,
    pub pci_irqs: [u32; 8],
}

static mut G_INFO: FdtInfo = FdtInfo {
    uart_base: 0,
    virtio_bases: [0; 32],
    virtio_irqs: [0; 32],
    virtio_count: 0,
    gic_dist_base: 0,
    gic_redist_base: 0,
    gic_version: 0,
    pcie_ecam_base: 0,
    pcie_ecam_size: 0,
    ram_base: 0,
    ram_size: 0,
    pci_mmio32_base: 0,
    pci_mmio32_size: 0,
    pci_irqs: [0; 8],
};

// ============================================================================
// Big-endian helpers
// ============================================================================

#[inline]
fn be32(p: *const u8) -> u32 {
    unsafe {
        ((*p.add(0) as u32) << 24)
            | ((*p.add(1) as u32) << 16)
            | ((*p.add(2) as u32) << 8)
            | (*p.add(3) as u32)
    }
}

#[inline]
fn be64_2cell(p: *const u8) -> u64 {
    ((be32(p) as u64) << 32) | be32(unsafe { p.add(4) }) as u64
}

// ============================================================================
// String helpers
// ============================================================================

fn str_eq(a: *const u8, b: &[u8]) -> bool {
    let mut i = 0;
    unsafe {
        while i < b.len() {
            if *a.add(i) != b[i] {
                return false;
            }
            i += 1;
        }
        // a must also be null-terminated here
        *a.add(i) == 0
    }
}

/// Check if the null-string list [data, data+len) contains needle.
fn strlist_has(data: *const u8, len: u32, needle: &[u8]) -> bool {
    let end = unsafe { data.add(len as usize) };
    let mut p = data;
    while (p as usize) < (end as usize) {
        // Check if current string matches needle
        let mut matches = true;
        let mut i = 0;
        let start = p;
        unsafe {
            while (p as usize) < (end as usize) && *p != 0 {
                if i < needle.len() {
                    if *p != needle[i] {
                        matches = false;
                    }
                }
                i += 1;
                p = p.add(1);
            }
            if matches && i == needle.len() {
                return true;
            }
            // Skip null terminator
            if (p as usize) < (end as usize) {
                p = p.add(1);
            }
        }
    }
    false
}

// ============================================================================
// DTB structure tokens
// ============================================================================

const FDT_MAGIC: u32 = 0xD00DFEED;
const FDT_BEGIN_NODE: u32 = 1;
const FDT_END_NODE: u32 = 2;
const FDT_PROP: u32 = 3;
const FDT_NOP: u32 = 4;
const FDT_END: u32 = 9;

// Compatible flags
const CF_PL011: u32 = 1 << 0;
const CF_VIRTIO: u32 = 1 << 1;
const CF_GICV2: u32 = 1 << 2;
const CF_PCI: u32 = 1 << 3;
const CF_MEMORY: u32 = 1 << 4;
const CF_GICV3: u32 = 1 << 5;

// ============================================================================
// Per-node state during parsing
// ============================================================================

const MAX_DEPTH: usize = 8;

struct NodeState {
    compat: u32,
    reg: u64,
    reg_size: u64,
    has_reg: bool,
    reg2: u64,
    has_reg2: bool,
    ranges_data: *const u8,
    ranges_len: u32,
    irq: u32,
    has_irq: bool,
    int_map_data: *const u8,
    int_map_len: u32,
}

impl NodeState {
    const fn new() -> Self {
        NodeState {
            compat: 0,
            reg: 0,
            reg_size: 0,
            has_reg: false,
            reg2: 0,
            has_reg2: false,
            ranges_data: ptr::null(),
            ranges_len: 0,
            irq: 0,
            has_irq: false,
            int_map_data: ptr::null(),
            int_map_len: 0,
        }
    }
}

// ============================================================================
// Main parser
// ============================================================================

#[cfg(target_arch = "aarch64")]
unsafe fn parse_dtb(dtb_addr: u64) {
    if dtb_addr == 0 {
        return;
    }
    let dtb = dtb_addr as *const u8;

    if be32(dtb) != FDT_MAGIC {
        return;
    }

    let off_struct = be32(dtb.add(8)) as usize;
    let off_strings = be32(dtb.add(12)) as usize;

    let strings = dtb.add(off_strings);
    let mut p = dtb.add(off_struct);

    let mut stack = [const { NodeState::new() }; MAX_DEPTH];
    let mut depth: i32 = 0;

    loop {
        let tok = be32(p);
        p = p.add(4);

        match tok {
            FDT_BEGIN_NODE => {
                if (depth as usize) < MAX_DEPTH {
                    stack[depth as usize] = NodeState::new();
                }
                depth += 1;
                // Skip null-terminated node name
                while *p != 0 {
                    p = p.add(1);
                }
                p = p.add(1); // skip null
                // Align to 4-byte boundary
                p = ((p as usize + 3) & !3) as *const u8;
            }

            FDT_END_NODE => {
                depth -= 1;
                if depth >= 0 && (depth as usize) < MAX_DEPTH {
                    let ns = &stack[depth as usize];
                    if ns.has_reg {
                        if (ns.compat & CF_PL011) != 0 && G_INFO.uart_base == 0 {
                            G_INFO.uart_base = ns.reg;
                        }

                        if (ns.compat & CF_VIRTIO) != 0 && G_INFO.virtio_count < 32 {
                            let idx = G_INFO.virtio_count as usize;
                            G_INFO.virtio_bases[idx] = ns.reg;
                            G_INFO.virtio_irqs[idx] = if ns.has_irq { ns.irq } else { 0 };
                            G_INFO.virtio_count += 1;
                        }

                        if (ns.compat & CF_GICV2) != 0 && G_INFO.gic_dist_base == 0 {
                            G_INFO.gic_dist_base = ns.reg;
                            G_INFO.gic_version = 2;
                        }

                        if (ns.compat & CF_GICV3) != 0 && G_INFO.gic_dist_base == 0 {
                            G_INFO.gic_dist_base = ns.reg;
                            G_INFO.gic_redist_base = if ns.has_reg2 { ns.reg2 } else { 0 };
                            G_INFO.gic_version = 3;
                        }

                        if (ns.compat & CF_PCI) != 0 && G_INFO.pcie_ecam_base == 0 {
                            G_INFO.pcie_ecam_base = ns.reg;
                            G_INFO.pcie_ecam_size = ns.reg_size;
                        }

                        // Parse PCI "ranges" for 32-bit MMIO aperture
                        if (ns.compat & CF_PCI) != 0
                            && !ns.ranges_data.is_null()
                            && ns.ranges_len >= 28
                            && G_INFO.pci_mmio32_base == 0
                        {
                            let mut off: u32 = 0;
                            while off + 28 <= ns.ranges_len {
                                let e = ns.ranges_data.add(off as usize);
                                let flags = be32(e);
                                let space = (flags >> 24) & 3;
                                if space == 2 {
                                    // 32-bit memory space
                                    let cpu_hi = be32(e.add(12)) as u64;
                                    let cpu_lo = be32(e.add(16)) as u64;
                                    G_INFO.pci_mmio32_base = (cpu_hi << 32) | cpu_lo;
                                    G_INFO.pci_mmio32_size =
                                        ((be32(e.add(20)) as u64) << 32) | be32(e.add(24)) as u64;
                                    break;
                                }
                                off += 28;
                            }
                        }

                        // Parse PCI interrupt-map
                        if (ns.compat & CF_PCI) != 0
                            && !ns.int_map_data.is_null()
                            && ns.int_map_len >= 40
                        {
                            let mut off: u32 = 0;
                            while off + 40 <= ns.int_map_len {
                                let e = ns.int_map_data.add(off as usize);
                                let child_hi = be32(e);
                                let gic_type = be32(e.add(28));
                                let gic_irq = be32(e.add(32));
                                let slot = (child_hi >> 11) & 0x1F;
                                let intid = if gic_type == 0 {
                                    gic_irq + 32
                                } else {
                                    gic_irq + 16
                                };
                                if slot < 8 && G_INFO.pci_irqs[slot as usize] == 0 {
                                    G_INFO.pci_irqs[slot as usize] = intid;
                                }
                                off += 40;
                            }
                        }

                        if (ns.compat & CF_MEMORY) != 0 && G_INFO.ram_size == 0 {
                            G_INFO.ram_base = ns.reg;
                            G_INFO.ram_size = ns.reg_size;
                        }
                    }
                }
            }

            FDT_PROP => {
                let vlen = be32(p);
                p = p.add(4);
                let nameoff = be32(p);
                p = p.add(4);
                let vdata = p;
                // Advance past value (padded to 4-byte alignment)
                p = p.add(((vlen + 3) & !3) as usize);

                let d = depth - 1;
                if d >= 0 && (d as usize) < MAX_DEPTH {
                    let ns = &mut stack[d as usize];
                    let pname = strings.add(nameoff as usize);

                    if str_eq(pname, b"compatible") {
                        if strlist_has(vdata, vlen, b"arm,pl011") {
                            ns.compat |= CF_PL011;
                        }
                        if strlist_has(vdata, vlen, b"virtio,mmio") {
                            ns.compat |= CF_VIRTIO;
                        }
                        if strlist_has(vdata, vlen, b"arm,gic-400")
                            || strlist_has(vdata, vlen, b"arm,cortex-a15-gic")
                        {
                            ns.compat |= CF_GICV2;
                        }
                        if strlist_has(vdata, vlen, b"pci-host-ecam-generic") {
                            ns.compat |= CF_PCI;
                        }
                        if strlist_has(vdata, vlen, b"arm,gic-v3") {
                            ns.compat |= CF_GICV3;
                        }
                    } else if str_eq(pname, b"device_type") {
                        if strlist_has(vdata, vlen, b"pci") {
                            ns.compat |= CF_PCI;
                        }
                        if strlist_has(vdata, vlen, b"memory") {
                            ns.compat |= CF_MEMORY;
                        }
                    } else if str_eq(pname, b"reg") && vlen >= 8 && !ns.has_reg {
                        ns.reg = be64_2cell(vdata);
                        ns.reg_size = if vlen >= 16 { be64_2cell(vdata.add(8)) } else { 0 };
                        ns.has_reg = true;
                        if vlen >= 32 {
                            ns.reg2 = be64_2cell(vdata.add(16));
                            ns.has_reg2 = true;
                        }
                    } else if str_eq(pname, b"ranges") && vlen > 0 {
                        ns.ranges_data = vdata;
                        ns.ranges_len = vlen;
                    } else if str_eq(pname, b"interrupt-map") {
                        ns.int_map_data = vdata;
                        ns.int_map_len = vlen;
                    } else if str_eq(pname, b"interrupts") && vlen >= 12 && !ns.has_irq {
                        let irq_type = be32(vdata);
                        let irq_num = be32(vdata.add(4));
                        ns.irq = if irq_type == 0 {
                            irq_num + 32
                        } else {
                            irq_num + 16
                        };
                        ns.has_irq = true;
                    }
                }
            }

            FDT_NOP => {}

            FDT_END => return,

            _ => return, // corrupted DTB
        }
    }
}

// ============================================================================
// Public Rust API — for Rust crate consumers (serial_rs, drivers_rs, etc.)
// ============================================================================

/// Return a reference to the global parsed FDT info.
///
/// # Safety
/// Caller must ensure `fdt_init()` has been called first.
pub unsafe fn info() -> &'static FdtInfo {
    &*(&raw const G_INFO)
}

// ============================================================================
// Public API
// ============================================================================

/// Parse the DTB at the given physical address.
/// Safe to call with dtb_addr=0 (no-op).
pub fn fdt_init(dtb_addr: u64) {
    #[cfg(target_arch = "aarch64")]
    unsafe { parse_dtb(dtb_addr); }
    #[cfg(not(target_arch = "aarch64"))]
    { let _ = dtb_addr; }
}

/// Return a pointer to the parsed FDT info.
pub fn fdt_info_ptr() -> *const FdtInfo {
    unsafe { &raw const G_INFO }
}

/// Copy the parsed FDT info to the caller's buffer.
pub fn fdt_get_info(out: *mut FdtInfo) {
    unsafe { ptr::copy_nonoverlapping(&raw const G_INFO, out, 1); }
}
